src/include/btree.i

```diff
@@ -1693,3 +1618,155 @@ __wt_split_descent_race(
+	 * starts out with the following page-index:
+	 *
+	 *	| ... | a | g | i | ... |
+	 *
+	 * The split page starts out with the following page-index:
+	 *
+	 *	| a | b | c | d | e | f |
+	 *
+	 * The first step is to move the c-f ranges into a new subtree, so, for
+	 * example we might have two new internal pages 'c' and 'e', where the
+	 * new 'c' page references the c-d namespace and the new 'e' page
+	 * references the e-f namespace. The top of the subtree references the
+	 * parent page, but until the parent's page index is updated, threads in
+	 * the subtree won't be able to ascend out of the subtree. However, once
+	 * the parent page's page index is updated to this:
+	 *
+	 *	| ... | a | c | e | g | i | ... |
+	 *
+	 * threads in the subtree can ascend into the parent. Imagine a cursor
+	 * in the c-d part of the namespace that ascends to the parent's 'c'
+	 * slot. It would then decrement to the slot before the 'c' slot, the
+	 * 'a' slot.
+	 *
+	 * The previous-cursor movement selects the last slot in the 'a' page;
+	 * if the split page's page-index hasn't been updated yet, it selects
+	 * the 'f' slot, which is incorrect. Once the split page's page index is
+	 * updated to this:
+	 *
+	 *	| a | b |
+	 *
+	 * the previous-cursor movement will select the 'b' slot, which is
+	 * correct.
+	 *
+	 * This function takes an argument which is the internal page into which
+	 * we're descending. If the last slot on the page no longer points to
+	 * the current page as its "home", the page is being split and part of
+	 * its namespace moved, we have to restart.
+	 */
+	WT_INTL_INDEX_GET(session, ref->page, pindex);
+	return (pindex->index[pindex->entries - 1]->home != ref->page);
+}
+
+/*
+ * __wt_page_swap_func --
+ *	Swap one page's hazard pointer for another one when hazard pointer
+ * coupling up/down the tree.
+ */
+static inline int
+__wt_page_swap_func(WT_SESSION_IMPL *session,
+    WT_REF *held, WT_REF *want, bool prev_race, uint32_t flags
+#ifdef HAVE_DIAGNOSTIC
+    , const char *file, int line
+#endif
+    )
+{
+	WT_DECL_RET;
+	bool acquired;
+
+	/*
+	 * This function is here to simplify the error handling during hazard
+	 * pointer coupling so we never leave a hazard pointer dangling.  The
+	 * assumption is we're holding a hazard pointer on "held", and want to
+	 * acquire a hazard pointer on "want", releasing the hazard pointer on
+	 * "held" when we're done.
+	 *
+	 * When walking the tree, we sometimes swap to the same page. Fast-path
+	 * that to avoid thinking about error handling.
+	 */
+	if (held == want)
+		return (0);
+
+	/* Get the wanted page. */
+	ret = __wt_page_in_func(session, want, flags
+#ifdef HAVE_DIAGNOSTIC
+	    , file, line
+#endif
+	    );
+
+	/*
+	 * Expected failures: page not found or restart. Our callers list the
+	 * errors they're expecting to handle.
+	 */
+	if (LF_ISSET(WT_READ_NOTFOUND_OK) && ret == WT_NOTFOUND)
+		return (WT_NOTFOUND);
+	if (LF_ISSET(WT_READ_RESTART_OK)) {
```

Comment:
Is this a proxy for any caller doing a search rather than a next/prev? i.e: callers from search rather than bt_walk? I'd rather the condition where the prev race is checked to be made explicit.

Reply:
I think what you mean is that we should always check `prev_race`, and if set (and we're looking at an internal page), call `__wt_split_prev_race`.  If we detect a race but `WT_READ_RESTART_OK` is not passed in, the error case at line 1767 should apply: https://github.com/wiredtiger/wiredtiger/pull/4011/commits/268d1c46d2df6dce5250a57fcab08143259b89bc#diff-b2442811cb1095a2949962a9b1fa9c24R1767

Is that the change you'd like?

Reply:
Actually, isn't the logic backwards in that error mapping code?

```
+	/*
+	 * If we're returning an error, don't let it be one our caller expects
+	 * to handle as returned by page-in: the expectation includes the held
+	 * page not having been released, and that's not the case.
+	 */
+	if (LF_ISSET(WT_READ_NOTFOUND_OK) && ret == WT_NOTFOUND)
+		WT_RET_MSG(session,
+		    EINVAL, "page-release WT_NOTFOUND error mapped to EINVAL");
+	if (LF_ISSET(WT_READ_RESTART_OK) && ret == WT_RESTART)
+		WT_RET_MSG(session,
+		    EINVAL, "page-release WT_RESTART error mapped to EINVAL");
```

They should be `if(!LF_ISSET(WT_READ_NOTFOUND_OK) && ret == WT_NOTFOUND) ...` etc. unless I'm missing something?

Reply:
Thanks for stating my concern more clearly Michael.

Reply:
@michaelcahill, @agorrod, fix pushed in 8ab5a47.

I also tightened down the conditions under which we check for a race, the race is only interesting when walking backwards through the tree.